File: android/src/main/kotlin/com/adyen/checkout/flutter/CheckoutPlatformApi.kt
package com.adyen.checkout.flutter

import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import com.adyen.checkout.core.action.data.Action
import com.adyen.checkout.core.action.data.ActionComponentData
import com.adyen.checkout.core.common.AdyenLogLevel
import com.adyen.checkout.core.common.AdyenLogger
import com.adyen.checkout.core.common.CheckoutContext
import com.adyen.checkout.core.components.ActionOnlyCheckoutCallbacks
import com.adyen.checkout.core.components.AdditionalDetailsResult
import com.adyen.checkout.core.components.AdvancedCheckoutResult
import com.adyen.checkout.core.components.Checkout
import com.adyen.checkout.core.components.CheckoutController
import com.adyen.checkout.core.common.CheckoutResultCode
import com.adyen.checkout.core.components.CheckoutAction
import com.adyen.checkout.core.components.data.model.paymentmethod.PaymentMethods
import com.adyen.checkout.core.error.CheckoutError
import com.adyen.checkout.cse.CardEncrypter
import com.adyen.checkout.flutter.apiOnly.AdyenCSE
import com.adyen.checkout.flutter.apiOnly.CardValidation
import com.adyen.checkout.flutter.components.CheckoutComponentRegistry
import com.adyen.checkout.flutter.components.ComponentPlatformEventHandler
import com.adyen.checkout.flutter.generated.ActionComponentDataDTO
import com.adyen.checkout.flutter.generated.ActionOnlyFlutterApi
import com.adyen.checkout.flutter.generated.AdyenPigeonError
import com.adyen.checkout.flutter.generated.AdvancedCheckoutResultDTO
import com.adyen.checkout.flutter.generated.CheckoutConfigurationDTO
import com.adyen.checkout.flutter.generated.CheckoutEventDTO
import com.adyen.checkout.flutter.generated.CheckoutHostApi
import com.adyen.checkout.flutter.generated.CheckoutSetupResultDTO
import com.adyen.checkout.flutter.generated.EncryptedCardDTO
import com.adyen.checkout.flutter.generated.SessionResponseDTO
import com.adyen.checkout.flutter.generated.UnencryptedCardDTO
import com.adyen.checkout.flutter.session.CheckoutHolder
import com.adyen.checkout.flutter.utils.ConfigurationMapper.toCheckoutConfiguration
import com.google.android.gms.wallet.WalletConstants
import com.adyen.threeds2.ThreeDS2Service
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import kotlin.coroutines.resume

internal class CheckoutPlatformApi(
    private val activity: FragmentActivity,
    private val checkoutHolder: CheckoutHolder,
    private val callbacksApi: com.adyen.checkout.flutter.generated.CheckoutCallbacksFlutterApi,
    private val actionOnlyApi: ActionOnlyFlutterApi,
    private val eventHandler: ComponentPlatformEventHandler,
) : CheckoutHostApi {
    private var actionController: CheckoutController? = null
    private var actionView: ComposeView? = null

    override fun setupSession(
        sessionResponse: SessionResponseDTO,
        configuration: CheckoutConfigurationDTO,
        callback: (Result<CheckoutSetupResultDTO>) -> Unit,
    ) {
        activity.lifecycleScope.launch {
            try {
                if (checkoutHolder.isActive()) {
                    throw AdyenPigeonError(
                        "CheckoutAlreadyActive",
                        "Another checkout flow is already active.",
                    )
                }
                val nativeConfiguration = configuration.toCheckoutConfiguration()
                val result = Checkout.setup(
                    sessionResponse = sessionResponse.toNativeSessionResponse(),
                    configuration = nativeConfiguration,
                )
                when (result) {
                    is Checkout.Result.Error -> throw result.error.toPigeonError()
                    is Checkout.Result.Success -> {
                        val checkoutId = UUID.randomUUID().toString()
                        checkoutHolder.put(checkoutId, result.checkoutContext)
                        callback(Result.success(result.checkoutContext.toSessionSetupResult(checkoutId)))
                    }
                }
            } catch (error: Throwable) {
                callback(Result.failure(error.toPigeonError()))
            }
        }
    }

    override fun setupAdvanced(
        paymentMethodsJson: String,
        configuration: CheckoutConfigurationDTO,
        callback: (Result<CheckoutSetupResultDTO>) -> Unit,
    ) {
        activity.lifecycleScope.launch {
            try {
                if (checkoutHolder.isActive()) {
                    throw AdyenPigeonError(
                        "CheckoutAlreadyActive",
                        "Another checkout flow is already active.",
                    )
                }
                val paymentMethods = PaymentMethods.SERIALIZER.deserialize(JSONObject(paymentMethodsJson))
                val result = Checkout.setup(
                    paymentMethods = paymentMethods,
                    configuration = configuration.toCheckoutConfiguration(),
                )
                when (result) {
                    is Checkout.Result.Error -> throw result.error.toPigeonError()
                    is Checkout.Result.Success -> {
                        val checkoutId = UUID.randomUUID().toString()
                        checkoutHolder.put(checkoutId, result.checkoutContext)
                        callback(Result.success(setupResult(checkoutId, paymentMethods)))
                    }
                }
            } catch (error: Throwable) {
                callback(Result.failure(error.toPigeonError()))
            }
        }
    }

    override fun disposeCheckout(checkoutId: String) {
        checkoutHolder.remove(checkoutId)
        CheckoutComponentRegistry.clearCheckout(checkoutId)
    }

    override fun handleAction(
        actionId: String,
        actionJson: String,
        configuration: CheckoutConfigurationDTO,
        callback: (Result<AdvancedCheckoutResultDTO>) -> Unit,
    ) {
        if (!checkoutHolder.beginAction(actionId)) {
            callback(
                Result.failure(
                    AdyenPigeonError(
                        "CheckoutAlreadyActive",
                        "Another checkout flow is already active.",
                    ),
                ),
            )
            return
        }

        activity.lifecycleScope.launch {
            try {
                val action = Action.SERIALIZER.deserialize(JSONObject(actionJson))
                val setupResult = Checkout.setup(
                    action = action,
                    configuration = configuration.toCheckoutConfiguration(),
                )
                val context = when (setupResult) {
                    is Checkout.Result.Error -> throw setupResult.error.toPigeonError()
                    is Checkout.Result.Success -> setupResult.checkoutContext
                }
                val result = CompletableDeferred<AdvancedCheckoutResult>()
                val callbacks = ActionOnlyCheckoutCallbacks(
                    onAdditionalDetails = { data ->
                        try {
                            requestActionAdditionalDetails(actionId, data)
                        } catch (error: Throwable) {
                            result.completeExceptionally(error.toPigeonError())
                            AdditionalDetailsResult.Completion(CheckoutResultCode.ERROR.value)
                        }
                    },
                    onFailure = { error ->
                        result.completeExceptionally(error.toPigeonError())
                    },
                    onComplete = { checkoutResult ->
                        result.complete(checkoutResult)
                    },
                )
                val controller = CheckoutController(
                    context = context,
                    callbacks = callbacks,
                    coroutineScope = activity.lifecycleScope,
                )
                actionController = controller
                showAction(controller)
                val checkoutResult = result.await()
                callback(Result.success(AdvancedCheckoutResultDTO(checkoutResult.resultCode.value)))
            } catch (error: Throwable) {
                callback(Result.failure(error.toPigeonError()))
            } finally {
                hideAction()
                actionController = null
                checkoutHolder.endAction(actionId)
            }
        }
    }

    override fun enableConsoleLogging(enabled: Boolean) {
        AdyenLogger.setLogLevel(if (enabled) AdyenLogLevel.VERBOSE else AdyenLogLevel.NONE)
    }

    override fun encryptCard(
        card: UnencryptedCardDTO,
        publicKey: String,
        callback: (Result<EncryptedCardDTO>) -> Unit,
    ) {
        try {
            callback(Result.success(AdyenCSE.encryptCard(card, publicKey)))
        } catch (error: Throwable) {
            callback(Result.failure(error.toPigeonError()))
        }
    }

    override fun encryptBin(
        bin: String,
        publicKey: String,
        callback: (Result<String>) -> Unit,
    ) {
        try {
            callback(Result.success(AdyenCSE.encryptBin(bin, publicKey)))
        } catch (error: Throwable) {
            callback(Result.failure(error.toPigeonError()))
        }
    }

    override fun validateCardNumber(cardNumber: String, enableLuhnCheck: Boolean): Boolean =
        CardValidation.validateCardNumber(cardNumber, enableLuhnCheck)

    override fun validateCardExpiryDate(expiryMonth: String, expiryYear: String): Boolean =
        CardValidation.validateCardExpiryDate(expiryMonth, expiryYear)

    override fun validateCardSecurityCode(securityCode: String, cardBrand: String?): Boolean =
        CardValidation.validateCardSecurityCode(securityCode, cardBrand)

    override fun getThreeDS2SdkVersion(): String = ThreeDS2Service.INSTANCE.sdkVersion

    fun handleReturn(intent: android.content.Intent) {
        actionController?.handleReturn(intent)
    }

    fun teardown() {
        hideAction()
        actionController = null
        checkoutHolder.clear()
    }

    private suspend fun requestActionAdditionalDetails(
        actionId: String,
        data: ActionComponentData,
    ): AdditionalDetailsResult = suspendCancellableCoroutine { continuation ->
        val dataJson = ActionComponentData.SERIALIZER.serialize(data).toString()
        actionOnlyApi.onAdditionalDetails(
            actionId,
            ActionComponentDataDTO(dataJson = dataJson),
        ) { response ->
            if (!continuation.isActive) return@onAdditionalDetails
            response.fold(
                onSuccess = { result ->
                    continuation.resume(AdditionalDetailsResult.Completion(result.resultCode))
                },
                onFailure = { error -> continuation.resumeWith(Result.failure(error)) },
            )
        }
        continuation.invokeOnCancellation { }
    }

    private fun showAction(controller: CheckoutController) {
        val view = ComposeView(activity).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnDetachedFromWindowOrReleasedFromPool)
            setContent {
                CheckoutAction(controller = controller)
            }
        }
        actionView = view
        activity.addContentView(
            view,
            FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            ),
        )
    }

    private fun hideAction() {
        actionView?.let { view ->
            (view.parent as? ViewGroup)?.removeView(view)
        }
        actionView = null
    }

    private fun SessionResponseDTO.toNativeSessionResponse(): com.adyen.checkout.core.sessions.SessionResponse =
        com.adyen.checkout.core.sessions.SessionResponse(id, sessionData)

    private fun CheckoutContext.Sessions.toSessionSetupResult(checkoutId: String): CheckoutSetupResultDTO {
        val methods = checkoutSession.sessionSetupResponse.paymentMethods
        return CheckoutSetupResultDTO(
            checkoutId = checkoutId,
            regularPaymentMethodsJson = serializeMethods(methods?.paymentMethods.orEmpty()),
            storedPaymentMethodsJson = serializeStoredMethods(methods?.storedPaymentMethods.orEmpty()),
        )
    }

    private fun setupResult(
        checkoutId: String,
        methods: PaymentMethods,
    ): CheckoutSetupResultDTO = CheckoutSetupResultDTO(
        checkoutId = checkoutId,
        regularPaymentMethodsJson = serializeMethods(methods.paymentMethods.orEmpty()),
        storedPaymentMethodsJson = serializeStoredMethods(methods.storedPaymentMethods.orEmpty()),
    )

    private fun serializeMethods(
        methods: List<com.adyen.checkout.core.components.data.model.paymentmethod.PaymentMethod>,
    ): String {
        val array = JSONArray()
        methods.forEach { method ->
            array.put(com.adyen.checkout.core.components.data.model.paymentmethod.PaymentMethod.SERIALIZER.serialize(method))
        }
        return array.toString()
    }

    private fun serializeStoredMethods(
        methods: List<com.adyen.checkout.core.components.data.model.paymentmethod.StoredPaymentMethod>,
    ): String {
        val array = JSONArray()
        methods.forEach { method ->
            array.put(com.adyen.checkout.core.components.data.model.paymentmethod.StoredPaymentMethod.SERIALIZER.serialize(method))
        }
        return array.toString()
    }

    private fun CheckoutError.toPigeonError(): AdyenPigeonError =
        AdyenPigeonError(code = code, message = message)

    private fun Throwable.toPigeonError(): AdyenPigeonError = when (this) {
        is AdyenPigeonError -> this
        else -> AdyenPigeonError(
            code = "Generic",
            message = message ?: "Checkout failed.",
        )
    }
}

File: ios/adyen_checkout/Sources/adyen_checkout/CheckoutPlatformApi.swift
import Adyen
import AdyenActions
import AdyenCard
import AdyenCheckout
import AdyenComponents
import AdyenEncryption
import AdyenSession
import Flutter
import Foundation
import UIKit

// swiftlint:disable type_body_length
@MainActor
final class CheckoutPlatformApi: CheckoutHostApi {
    private let callbacksApi: CheckoutCallbacksFlutterApi
    private let actionOnlyApi: ActionOnlyFlutterApi
    private let holder: CheckoutHolder
    private let events: ComponentPlatformEventHandler
    private var presentationDelegates: [String: PresentationDelegateProxy] = [:]
    private var actionCompletion: AsyncResult<AdvancedCheckoutResult>?

    init(
        callbacksApi: CheckoutCallbacksFlutterApi,
        actionOnlyApi: ActionOnlyFlutterApi,
        holder: CheckoutHolder,
        events: ComponentPlatformEventHandler
    ) {
        self.callbacksApi = callbacksApi
        self.actionOnlyApi = actionOnlyApi
        self.holder = holder
        self.events = events
    }

    func setupSession(
        sessionResponse: SessionResponseDTO,
        configuration: CheckoutConfigurationDTO,
        completion: @escaping (Result<CheckoutSetupResultDTO, Error>) -> Void
    ) {
        let checkoutId = UUID().uuidString
        Task { @MainActor in
            do {
                guard !holder.isActive else {
                    throw AdyenPigeonError(code: "CheckoutAlreadyActive", message: "Another checkout flow is already active.", details: nil)
                }
                let presentationDelegate = makePresentationDelegate(id: checkoutId)
                let checkout = try await Checkout.setup(
                    with: sessionResponse.toSessionResponse(),
                    configuration: configuration.toCheckoutConfiguration(callbacksApi: callbacksApi, checkoutId: checkoutId),
                    presentationDelegate: presentationDelegate
                )
                .onBeforeSubmit { [weak self] data in
                    guard let self else { return .abort }
                    return await self.requestBeforeSubmit(checkoutId: checkoutId, data: data)
                }
                .onComplete { [weak self] result in
                    self?.sendComplete(checkoutId: checkoutId, result: result)
                }
                .onFailure { [weak self] error in
                    self?.sendFailure(checkoutId: checkoutId, error: error)
                }
                holder.store(checkout, id: checkoutId)
                try completion(.success(checkout.setupResult(checkoutId: checkoutId)))
            } catch {
                completion(.failure(asPigeonError(error)))
            }
        }
    }

    func setupAdvanced(
        paymentMethodsJson: String,
        configuration: CheckoutConfigurationDTO,
        completion: @escaping (Result<CheckoutSetupResultDTO, Error>) -> Void
    ) {
        let checkoutId = UUID().uuidString
        Task { @MainActor in
            do {
                guard !holder.isActive else {
                    throw AdyenPigeonError(code: "CheckoutAlreadyActive", message: "Another checkout flow is already active.", details: nil)
                }
                let paymentMethods = try JSONDecoder().decode(PaymentMethods.self, from: Data(paymentMethodsJson.utf8))
                let presentationDelegate = makePresentationDelegate(id: checkoutId)
                let checkout = try await Checkout.setup(
                    with: paymentMethods,
                    configuration: configuration.toCheckoutConfiguration(callbacksApi: callbacksApi, checkoutId: checkoutId),
                    presentationDelegate: presentationDelegate
                )
                .onSubmit { [weak self] data in
                    guard let self else { return .retry(errorMessage: "Checkout is no longer active.") }
                    return await self.requestSubmit(checkoutId: checkoutId, data: data)
                }
                .onAdditionalDetails { [weak self] data in
                    guard let self else { return .completion(resultCode: "Error") }
                    return await self.requestAdditionalDetails(checkoutId: checkoutId, data: data)
                }
                .onComplete { [weak self] result in
                    self?.sendComplete(checkoutId: checkoutId, result: result)
                }
                .onFailure { [weak self] error in
                    self?.sendFailure(checkoutId: checkoutId, error: error)
                }
                holder.store(checkout, id: checkoutId)
                try completion(.success(checkout.setupResult(checkoutId: checkoutId)))
            } catch {
                completion(.failure(asPigeonError(error)))
            }
        }
    }

    func disposeCheckout(checkoutId: String) throws {
        holder.removeCheckout(for: checkoutId)
        presentationDelegates.removeValue(forKey: checkoutId)
        CheckoutComponentRegistry.shared.clear(checkoutId: checkoutId)
    }

    func handleAction(
        actionId: String,
        actionJson: String,
        configuration: CheckoutConfigurationDTO,
        completion: @escaping (Result<AdvancedCheckoutResultDTO, Error>) -> Void
    ) {
        guard holder.beginAction(id: actionId) else {
            completion(.failure(AdyenPigeonError(
                code: "CheckoutAlreadyActive",
                message: "Another checkout flow is already active.",
                details: nil
            )))
            return
        }
        Task { @MainActor in
            defer { holder.endAction(id: actionId) }
            do {
                let action = try JSONDecoder().decode(Action.self, from: Data(actionJson.utf8))
                let presentationDelegate = makePresentationDelegate(id: actionId)
                let checkout = try await Checkout.setup(
                    configuration: configuration.toCheckoutConfiguration(),
                    presentationDelegate: presentationDelegate
                )
                let result = AsyncResult<AdvancedCheckoutResult>()
                actionCompletion = result
                _ = checkout
                    .onAdditionalDetails { [weak self] data in
                        guard let self else { return .completion(resultCode: "Error") }
                        do {
                            return try await self.requestActionDetails(actionId: actionId, data: data)
                        } catch {
                            result.fail(error)
                            return .completion(resultCode: "Error")
                        }
                    }
                    .onComplete { value in result.succeed(value) }
                    .onFailure { error in result.fail(error) }
                holder.actionCheckout = checkout
                checkout.handle(action: action)
                let value = try await result.wait()
                completion(.success(AdvancedCheckoutResultDTO(resultCode: value.resultCode.rawValue)))
            } catch {
                completion(.failure(asPigeonError(error)))
            }
            actionCompletion = nil
            presentationDelegates.removeValue(forKey: actionId)
        }
    }

    func enableConsoleLogging(enabled: Bool) throws {
        AdyenLogging.isEnabled = enabled
    }

    func encryptCard(
        card: UnencryptedCardDTO,
        publicKey: String,
        completion: @escaping (Result<EncryptedCardDTO, Error>) -> Void
    ) {
        do {
            let encrypted = try CardEncryptor.encrypt(card: card.toCard(), with: publicKey)
            completion(.success(encrypted.toDTO()))
        } catch {
            completion(.failure(asPigeonError(error)))
        }
    }

    func encryptBin(
        bin: String,
        publicKey: String,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        do {
            try completion(.success(CardEncryptor.encrypt(bin: bin, with: publicKey)))
        } catch {
            completion(.failure(asPigeonError(error)))
        }
    }

    func validateCardNumber(cardNumber: String, enableLuhnCheck: Bool) throws -> Bool {
        CardValidation().validateCardNumber(cardNumber: cardNumber, enableLuhnCheck: enableLuhnCheck)
    }

    func validateCardExpiryDate(expiryMonth: String, expiryYear: String) throws -> Bool {
        CardValidation().validateCardExpiryDate(expiryMonth: expiryMonth, expiryYear: expiryYear)
    }

    func validateCardSecurityCode(securityCode: String, cardBrand: String?) throws -> Bool {
        CardValidation().validateCardSecurityCode(securityCode: securityCode, cardBrand: cardBrand)
    }

    func getThreeDS2SdkVersion() throws -> String {
        threeDS2SdkVersion
    }

    func handleReturn(url: URL) {
        _ = Checkout.handleReturn(url: url)
    }

    func teardown() {
        holder.clear()
        CheckoutComponentRegistry.shared.clear()
        presentationDelegates.removeAll()
        actionCompletion = nil
    }

    private func requestBeforeSubmit(checkoutId: String, data: BeforeSubmitData) async -> BeforeSubmitResult {
        await withCheckedContinuation { continuation in
            callbacksApi.onBeforeSubmit(
                checkoutId: checkoutId,
                data: data.toDTO()
            ) { result in
                switch result {
                case let .success(value):
                    continuation.resume(returning: value.toNativeResult(original: data))
                case let .failure(error):
                    self.sendFailure(checkoutId: checkoutId, code: "CallbackFailure", message: error.localizedDescription)
                    continuation.resume(returning: .abort)
                }
            }
        }
    }

    private func requestSubmit(checkoutId: String, data: PaymentComponentData) async -> SubmitResult {
        do {
            let dataJson = try EncodablePaymentComponentData(data: data).jsonString
            return await withCheckedContinuation { continuation in
                callbacksApi.onSubmit(
                    checkoutId: checkoutId,
                    data: PaymentComponentDataDTO(dataJson: dataJson)
                ) { result in
                    switch result {
                    case let .success(value):
                        do {
                            try continuation.resume(returning: value.toNativeResult())
                        } catch {
                            continuation.resume(returning: .retry(errorMessage: error.localizedDescription))
                        }
                    case let .failure(error):
                        continuation.resume(returning: .retry(errorMessage: error.localizedDescription))
                    }
                }
            }
        } catch {
            return .retry(errorMessage: error.localizedDescription)
        }
    }

    private func requestAdditionalDetails(checkoutId: String, data: ActionComponentData) async -> AdditionalDetailsResult {
        do {
            let dataJson = try EncodableActionComponentData(data: data).jsonString
            return await withCheckedContinuation { continuation in
                callbacksApi.onAdditionalDetails(
                    checkoutId: checkoutId,
                    data: ActionComponentDataDTO(dataJson: dataJson)
                ) { result in
                    switch result {
                    case let .success(value):
                        continuation.resume(returning: .completion(resultCode: value.resultCode))
                    case let .failure(error):
                        self.sendFailure(checkoutId: checkoutId, code: "CallbackFailure", message: error.localizedDescription)
                        continuation.resume(returning: .completion(resultCode: "Error"))
                    }
                }
            }
        } catch {
            sendFailure(checkoutId: checkoutId, code: "CallbackFailure", message: error.localizedDescription)
            return .completion(resultCode: "Error")
        }
    }

    private func requestActionDetails(actionId: String, data: ActionComponentData) async throws -> AdditionalDetailsResult {
        let dataJson = try EncodableActionComponentData(data: data).jsonString
        return try await withCheckedThrowingContinuation { continuation in
            actionOnlyApi.onAdditionalDetails(
                actionId: actionId,
                data: ActionComponentDataDTO(dataJson: dataJson)
            ) { result in
                switch result {
                case let .success(value):
                    continuation.resume(returning: .completion(resultCode: value.resultCode))
                case let .failure(error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    private func sendComplete(checkoutId: String, result: SessionCheckoutResult) {
        events.send(event: CheckoutEventDTO(
            type: .complete,
            checkoutId: checkoutId,
            resultCode: result.resultCode.rawValue,
            sessionId: result.sessionId,
            sessionData: result.sessionResult
        ))
    }

    private func sendComplete(checkoutId: String, result: AdvancedCheckoutResult) {
        events.send(event: CheckoutEventDTO(
            type: .complete,
            checkoutId: checkoutId,
            resultCode: result.resultCode.rawValue
        ))
    }

    private func sendFailure(checkoutId: String, error: CheckoutError) {
        sendFailure(checkoutId: checkoutId, code: error.code.rawValue, message: error.message)
    }

    private func sendFailure(checkoutId: String, code: String, message: String?) {
        events.send(event: CheckoutEventDTO(
            type: .failure,
            checkoutId: checkoutId,
            errorCode: code,
            errorMessage: message
        ))
    }

    private func makePresentationDelegate(id: String) -> PresentationDelegateProxy {
        if let delegate = presentationDelegates[id] {
            return delegate
        }
        let delegate = PresentationDelegateProxy(viewController: rootViewController())
        presentationDelegates[id] = delegate
        return delegate
    }

    private func rootViewController() -> UIViewController? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: { $0.isKeyWindow })?.rootViewController
    }
}

// swiftlint:enable type_body_length

@MainActor
private final class PresentationDelegateProxy: NSObject, Adyen.PresentationDelegate {
    weak var presentingViewController: UIViewController?

    init(viewController: UIViewController?) {
        self.presentingViewController = viewController
    }

    func present(component: PresentableComponent) {
        presentingViewController?.present(component.viewController, animated: true)
    }
}

private final class AsyncResult<Value> {
    private var result: Result<Value, Error>?
    private var continuation: CheckedContinuation<Value, Error>?

    func wait() async throws -> Value {
        if let result {
            return try result.get()
        }
        return try await withCheckedThrowingContinuation { continuation in
            self.continuation = continuation
        }
    }

    func succeed(_ value: Value) {
        resolve(.success(value))
    }

    func fail(_ error: Error) {
        resolve(.failure(error))
    }

    private func resolve(_ result: Result<Value, Error>) {
        guard self.result == nil else { return }
        self.result = result
        continuation?.resume(with: result)
        continuation = nil
    }
}

private func asPigeonError(_ error: Error) -> AdyenPigeonError {
    if let error = error as? AdyenPigeonError {
        return error
    }
    if let error = error as? CheckoutError {
        return AdyenPigeonError(code: error.code.rawValue, message: error.message, details: nil)
    }
    return AdyenPigeonError(code: "Generic", message: error.localizedDescription, details: nil)
}
